### PR TITLE
Add way to access mesh refinement strategies

### PR DESCRIPTION
**Pull request by class4kayaker**
Fix GH issue #2246

### CODE REVIEW

**Comment by bangerth** [comment]
OK to merge once the tester is happy.

File: include/aspect/mesh_refinement/interface.h [hunk 1]
```diff
@@ -222,6 +222,18 @@ namespace aspect
...
+         * Go through the list of all mesh refinement strategies that have been
+         * selected in the input file (and are consequently currently active)
+         * and see if one of them has the desired type specified by the
+         * template argument. If so, return a pointer to it. If no mesh
+         * refinement strategy is active that matches the given type, return a
+         * NULL pointer.
+         */
+        template <typename MeshRefinementType>
+        MeshRefinementType *
+        find_mesh_refinement_strategy () const;
```
Comment (gassmoeller) [inline include/aspect/mesh_refinement/interface.h]: I would prefer if we would avoid this interface, and instead use an approach similar to the new `has_matching_postprocessor` and `get_matching_postprocessor` functions in the postprocess/interface.h file. I do not like returning null pointers, and we deprecated this interface for the postprocessors.

**Comment by gassmoeller** [comment]
Oh and could you please rebase to the current master? The test failure should be fixed there.

**Comment by class4kayaker** [comment]
Changed interface as suggested. Awaiting test results.

File: include/aspect/mesh_refinement/interface.h [hunk 2]
```diff
@@ -301,6 +323,48 @@ namespace aspect
...
+
+
+    template <int dim>
+    template <typename MeshRefinementType>
+    inline
+    const MeshRefinementType &
+    Manager<dim>::get_matching_mesh_refinement_strategy () const
+    {
+      AssertThrow(has_matching_mesh_refinement_strategy<MeshRefinementType> (),
+                  ExcMessage("You asked MeshRefinement:Manager::get_matching_postprocessor() for a "
```
Comment (gassmoeller) [inline include/aspect/mesh_refinement/interface.h]: copy-paste leftover

File: include/aspect/mesh_refinement/interface.h [hunk 3]
```diff
@@ -301,6 +323,48 @@ namespace aspect
...
+                             "mesh refinement strategy of type <" + boost::core::demangle(typeid(MeshRefinementType).name()) + "> "
+                             "that could not be found in the current model. Activate this "
+                             "postprocessor in the input file."));
```
Comment (gassmoeller) [inline include/aspect/mesh_refinement/interface.h]: same here

File: include/aspect/mesh_refinement/interface.h [hunk 4]
```diff
@@ -301,6 +323,48 @@ namespace aspect
...
+
+      typename std::vector<std_cxx11::shared_ptr<Interface<dim> > >::const_iterator postprocessor;
```
Comment (gassmoeller) [inline include/aspect/mesh_refinement/interface.h]: and here

File: include/aspect/mesh_refinement/interface.h [hunk 5]
```diff
@@ -301,6 +323,48 @@ namespace aspect
...
+      for (typename std::vector<std_cxx11::shared_ptr<Interface<dim> > >::const_iterator
+           p = mesh_refinement_objects.begin();
+           p != mesh_refinement_objects.end(); ++p)
+        if (Plugins::plugin_type_matches<MeshRefinementType>(*(*p)))
+          return Plugins::get_plugin_as_type<MeshRefinementType>(*(*p));
+
+      // We will never get here, because we had the Assert above. Just to avoid warnings.
+      return Plugins::get_plugin_as_type<MeshRefinementType>(*(*postprocessor));
```
Comment (gassmoeller) [inline include/aspect/mesh_refinement/interface.h]: and here :smile:

**Comment by gassmoeller** [comment]
Yes, I was planning to go through all interfaces and deprecate the `find_` functions in favor of the `has_` and `get_` functions, but I never got to it. Something for future cleanup (any PR is appreciated :smile:).

**Comment by class4kayaker** [comment]
It might be worth opening it as an issue to be available as a minor hackathon project.

**Comment by gassmoeller** [comment]
Done see #2260.

**Comment by gassmoeller** [comment]
Thanks, looks good now. 
@tjhei: I do not have permissions to restart this PR on the jenkins tester with `TRUST_BUILD==true`. I guess we can note that the permissions check works, but the people who need to restart these tests need the permission to actually do so :smile:. Since the old tester passes I am going to merge.

**Pull request merged**